src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java

```diff
@@ -0,0 +1,200 @@
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiWhiteSpace;
+import com.intellij.psi.tree.IElementType;
+import com.intellij.psi.util.PsiTreeUtil;
+import io.ballerina.plugins.idea.psi.BallerinaDocumentationString;
+import io.ballerina.plugins.idea.psi.BallerinaFile;
+import io.ballerina.plugins.idea.psi.BallerinaImportDecl;
+import io.ballerina.plugins.idea.psi.BallerinaTypes;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
+
+/**
+ * Folding builder for Ballerina language.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaFoldingBuilder extends CustomFoldingBuilder implements DumbAware {
+
+    private static final String DEFAULT_PLACEHOLDER = "...";
+    private static final String COMMENT_PLACEHOLDER = "// ...";
+    private static final String DOCUMENTATION_PLACEHOLDER = "# ...";
+    private static final String BRACE_PLACEHOLDER = "{...}";
+    private static final String BRACE_PIPE_PLACEHOLDER = "{|...|}";
+
+    @Override
+    protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> list, @NotNull PsiElement psiElement,
+                                            @NotNull Document document, boolean b) {
+        if (!(psiElement instanceof BallerinaFile)) {
+            return;
+        }
+        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.OPEN_BRACE_TOKEN,
+                BallerinaTypes.CLOSE_BRACE_TOKEN, BRACE_PLACEHOLDER));
+        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.OPEN_NESTED_BRACE_TOKEN,
+                BallerinaTypes.CLOSE_NESTED_BRACE_TOKEN, BRACE_PLACEHOLDER));
+        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.IGNORED_OPEN_BRACE_TOKEN,
+                BallerinaTypes.IGNORED_CLOSE_BRACE_TOKEN, BRACE_PLACEHOLDER));
+        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.OPEN_BRACE_PIPE_TOKEN,
+                BallerinaTypes.CLOSE_BRACE_PIPE_TOKEN, BRACE_PIPE_PLACEHOLDER));
+        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.OPEN_NESTED_BRACE_PIPE_TOKEN,
+                BallerinaTypes.CLOSE_NESTED_BRACE_PIPE_TOKEN, BRACE_PIPE_PLACEHOLDER));
+        list.addAll(createImportFoldingRegions(psiElement));
+        list.addAll(createDocumentationFoldingRegions(psiElement));
+        list.addAll(createMultiCommentFoldingRegions(psiElement));
+    }
+
+    @Override
+    protected String getLanguagePlaceholderText(@NotNull ASTNode astNode, @NotNull TextRange textRange) {
+        return DEFAULT_PLACEHOLDER;
+    }
+
+    @Override
+    protected boolean isRegionCollapsedByDefault(@NotNull ASTNode astNode) {
+        return false;
+    }
+
+    private List<FoldingDescriptor> createFoldingRegions(PsiElement psiElement, IElementType open, IElementType close,
+                                      String placeholder) {
+        List<FoldingDescriptor> foldingDescriptors = new ArrayList<>();
```

Comment:
IMO its better to use linked lists instead, as here we need fast insertion and random access is not a primary concern.
```suggestion
        List<FoldingDescriptor> foldingDescriptors = new LinkedList<>();
```

Reply:
Please apply where neccessary